### PR TITLE
Add AbortController to the sampling profiler

### PR DESCRIPTION
**Pull request by mcollina**

### CODE REVIEW

File: src/preloader.js [hunk 1]
```diff
@@ -3,6 +3,7 @@
 const generateHeapSnapshot = require('./snapshot')
 const generateHeapSamplingProfile = require('./profile')
 const recordAllocationTimeline = require('./timeline')
+const AbortController = require('abort-controller')
```
Comment (jasnell) [inline src/preloader.js:6]: Should set this up only to use `require('abort-controller')` if the built-in `AbortController` added in 15.0.0 is not present.

File: src/preloader.js [hunk 2]
```diff
@@ -88,7 +89,15 @@ module.exports = function installPreloader(logger) {
     }
 
     if (takeProfile) {
-      benchmarkGeneration(logger, 'sampling profile', generateHeapSamplingProfile, profilerOptions, onToolEnd)
+      const controller = new AbortController()
+      profilerOptions.signal = controller.signal
+      const abort = controller.abort.bind(controller)
```
Comment (jasnell) [inline src/preloader.js:94]: Be sure to check to see if the `AbortSignal` has already been tripped before proceeding.

File: src/preloader.js [hunk 3]
```diff
@@ -88,7 +89,15 @@ module.exports = function installPreloader(logger) {
...
+      process.on('SIGUSR2', abort)
```
Comment (jasnell) [inline src/preloader.js:95]: The abort will be a non-op after the first time that it is triggered.

File: src/profile.js [hunk 4]
```diff
@@ -34,6 +35,35 @@ module.exports = function generateHeapSamplingProfile(options, cb) {
     throw new Error('The interval option must be a number greater than 0')
   }
 
+  if (signal) {
+    if (signal.addEventListener) {
```
Comment (jasnell) [inline src/profile.js:39]: Check that signal has not already been tripped here.

File: src/profile.js [hunk 5]
```diff
@@ -34,6 +35,35 @@ module.exports = function generateHeapSamplingProfile(options, cb) {
...
+      signal.addEventListener('abort', finish)
+    } else {
+      signal.on('abort', finish)
```
Comment (jasnell) [inline src/profile.js:42]: The abort should only ever trigger once.

**Review by ShogunPanda — APPROVED** [review]
Other than James' remarks, rest LGTM!

File: src/preloader.js (hunk 1 above)
Comment (mcollina) [reply src/preloader.js:6]: The added in abort-controller is emitting an experimental warning, I would prefer _not_ to use it.

File: src/preloader.js (hunk 2 above)
Comment (mcollina) [reply src/preloader.js:94]: done

File: src/profile.js (hunk 4 above)
Comment (mcollina) [reply src/profile.js:43]: done

**Comment by mcollina** [comment]
@jasnell PTAL

File: src/preloader.js (hunk 1 above)
Comment (jasnell) [reply src/preloader.js:6]: Ok. I'll be opening a PR soon to graduate that from experimental soon.

**Review by jasnell — APPROVED** [review]

**Pull request merged**